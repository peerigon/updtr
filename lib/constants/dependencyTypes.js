"use strict";

const REGULAR = 0;
const DEV = 1;
const OPTIONAL = 2;

function fromNpm(type) {
    switch (type) {
        case "dependencies":
            return REGULAR;
        case "devDependencies":
            return DEV;
        case "optionalDependencies":
            return OPTIONAL;
        default:
            throw new Error(`Unsupported dependency type: ${ type }`);
    }
}

exports.REGULAR = REGULAR;
exports.DEV = DEV;
exports.OPTIONAL = OPTIONAL;

exports.fromNpm = fromNpm;
// yarn uses the same strings. Change this if necessary.
exports.fromYarn = fromNpm;
