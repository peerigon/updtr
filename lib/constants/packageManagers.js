"use strict";

const NPM = "npm";
const YARN = "yarn";
const SUPPORTED = [NPM, YARN];

exports.NPM = NPM;
exports.YARN = YARN;
exports.SUPPORTED = SUPPORTED;
