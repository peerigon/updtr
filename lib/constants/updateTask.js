"use strict";

const WANTED = "wanted";
const LATEST = "latest";

exports.WANTED = WANTED;
exports.LATEST = LATEST;
