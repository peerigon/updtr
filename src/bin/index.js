import argv from "./argv";
import { create, run } from "./..";
import reporters from "../reporters";

async function start() {
    const cwd = process.cwd();
    const config = { ...argv };
    const reporterConfig = {
        stream: process.stdout,
        testStdout: argv.testStdout,
    };
    const reporter = reporters[argv.reporter];

    config.cwd = cwd;

    const updtr = create(config);

    reporter(updtr, reporterConfig);
    try {
        await run(updtr);
    } catch (err) {
        updtr.emit("error", err);
    }
}

start();
