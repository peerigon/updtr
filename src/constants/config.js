// The first value in arrays is the default value

export const USE_NPM = "npm";
export const USE_YARN = "yarn";
export const USE_OPTIONS = [USE_NPM, USE_YARN];
export const UPDATE_TO_LATEST = "latest";
export const UPDATE_TO_NON_BREAKING = "non-breaking";
export const UPDATE_TO_WANTED = "wanted";
export const UPDATE_TO_OPTIONS = [
    UPDATE_TO_LATEST,
    UPDATE_TO_NON_BREAKING,
    UPDATE_TO_WANTED,
];
export const SAVE_SMART = "smart";
export const SAVE_CARET = "caret";
export const SAVE_EXACT = "exact";
export const SAVE_OPTIONS = [SAVE_SMART, SAVE_CARET, SAVE_EXACT];
