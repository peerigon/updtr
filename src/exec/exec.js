import childProcess from "child_process";

function promiseExec(cwd, cmd) {
    return new Promise(resolve => {
        childProcess.exec(
            cmd,
            { maxBuffer: Infinity, encoding: "utf8", cwd },
            (err, stdout, stderr) => void resolve({ err, stdout, stderr })
        );
    });
}

export default (async function exec(cwd, cmd) {
    const { err, stdout, stderr } = await promiseExec(cwd, cmd);

    if (err !== null) {
        err.stdout = stdout;
        err.stderr = stderr;

        throw err;
    }

    return { stdout, stderr };
});
