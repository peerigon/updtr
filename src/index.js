import run from "./run";
import Updtr from "./Updtr";
import * as errors from "./errors";

export function create(config) {
    return new Updtr(config);
}

export { errors, run };
