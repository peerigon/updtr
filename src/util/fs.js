import fs from "fs";
import pify from "pify";

export default pify(fs);
