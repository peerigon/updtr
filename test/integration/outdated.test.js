import {
    USE_YARN,
    UPDATE_TO_NON_BREAKING,
    SAVE_CARET,
    SAVE_EXACT,
} from "../../src/constants/config";
import exec from "../../src/exec/exec";
import {
    outdatedRegular as outdatedPackageContents,
} from "../fixtures/packageJsons";
import { create, run } from "../../src";
import getInstalledVersions, {
    filterUpdtrTestModule,
} from "./helpers/getInstalledVersions";
import { PACKAGE_JSON, YARN_LOCK } from "./helpers/constants";
import { createTempDir, read, write } from "./helpers/fs";

// These tests may take longer on travis
jasmine.DEFAULT_TIMEOUT_INTERVAL = 60000;

describe("integration test: when there are outdated dependencies", () => {
    describe("default configuration", () => {
        describe("npm", () => {
            it("should install all updates, update the package.json and return the results", async () => {
                const tempDir = await createTempDir();

                await write(tempDir, PACKAGE_JSON, outdatedPackageContents);

                const updtr = create({
                    cwd: tempDir,
                });
                const results = await run(updtr);
                const [
                    tempDirPackageContents,
                    installedVersions,
                ] = await Promise.all([
                    read(tempDir, PACKAGE_JSON),
                    getInstalledVersions(tempDir, filterUpdtrTestModule),
                ]);

                expect(results).toMatchSnapshot(
                    "default configuration > npm > results"
                );
                expect(tempDirPackageContents).not.toBe(
                    outdatedPackageContents
                );
                expect(
                    JSON.parse(tempDirPackageContents).dependencies
                ).toMatchSnapshot(
                    "default configuration > npm > package.json dependencies"
                );
                expect(installedVersions.versions).toMatchSnapshot(
                    "default configuration > npm > installed versions"
                );
            });
        });
        describe("yarn", () => {
            it("should install all updates, update the package.json and return the results", async () => {
                const tempDir = await createTempDir();

                await write(tempDir, PACKAGE_JSON, outdatedPackageContents);

                const updtr = create({
                    cwd: tempDir,
                    use: USE_YARN,
                });
                const results = await run(updtr);
                const [
                    tempDirPackageContents,
                    installedVersions,
                    tempDirYarnLockContents,
                ] = await Promise.all([
                    read(tempDir, PACKAGE_JSON),
                    getInstalledVersions(tempDir, filterUpdtrTestModule),
                    read(tempDir, YARN_LOCK),
                ]);

                expect(results).toMatchSnapshot(
                    "default configuration > yarn > results"
                );
                expect(tempDirPackageContents).not.toBe(
                    outdatedPackageContents
                );
                expect(
                    JSON.parse(tempDirPackageContents).dependencies
                ).toMatchSnapshot(
                    "default configuration > yarn > package.json dependencies"
                );
                expect(installedVersions.versions).toMatchSnapshot(
                    "default configuration > yarn > installed versions"
                );
                expect(tempDirYarnLockContents).toMatch(
                    /updtr-test-module-1@2\.0\.0:\s*version "2\.0\.0"/
                );
                expect(tempDirYarnLockContents).toMatch(
                    // This module still has its caret because the latest version is
                    // already installed during the init task
                    /updtr-test-module-2@\^2\.0\.0:\s*version "2\.1\.1"/
                );
            });
        });
    });
    describe("advanced configuration", () => {
        describe("npm", () => {
            it("should install all updates, update the package.json and return the results", async () => {
                const tempDir = await createTempDir();
                const registry = "http://registry.npmjs.org";
                const expectedTestStdout = "This is a custom test command";
                let installMissingCmd;
                let testStdout;

                await write(tempDir, PACKAGE_JSON, outdatedPackageContents);
                // This time, we install all dependencies
                await exec(
                    tempDir,
                    "npm i updtr-test-module-1@1.0.0 updtr-test-module-2@2.0.0"
                );

                const installedVersionsBefore = getInstalledVersions(
                    tempDir,
                    filterUpdtrTestModule
                );
                const updtr = create({
                    cwd: tempDir,
                    exclude: ["updtr-test-module-2"],
                    updateTo: UPDATE_TO_NON_BREAKING,
                    test: `echo ${ expectedTestStdout } && exit 0`,
                    save: SAVE_EXACT,
                    registry,
                });

                updtr.on("init/install-missing", ({ cmd }) => {
                    installMissingCmd = cmd;
                });
                updtr.on("sequential-update/result", ({ stdout }) => {
                    // echo adds an EOL
                    testStdout = stdout.trim();
                });

                const results = await run(updtr);
                const [
                    tempDirPackageContents,
                    installedVersionsAfter,
                ] = await Promise.all([
                    read(tempDir, PACKAGE_JSON),
                    getInstalledVersions(tempDir, filterUpdtrTestModule),
                ]);

                expect(installMissingCmd).toContain(registry);
                expect(testStdout).toBe(expectedTestStdout);
                expect(results).toMatchSnapshot(
                    "advanced configuration > npm > results"
                );
                expect(tempDirPackageContents).not.toBe(
                    outdatedPackageContents
                );
                expect(
                    JSON.parse(tempDirPackageContents).dependencies
                ).toMatchSnapshot(
                    "advanced configuration > npm > package.json dependencies"
                );
                expect(installedVersionsAfter).not.toEqual(
                    installedVersionsBefore
                );
                expect(installedVersionsAfter.versions).toMatchSnapshot(
                    "advanced configuration > npm > installed versions"
                );
            });
        });
        describe("yarn", () => {
            it("should install all updates, update the package.json and return the results", async () => {
                // yarn does not support custom registries yet, thus it's commented out here.
                const tempDir = await createTempDir();
                // const registry = "http://registry.npmjs.org";
                const expectedTestStdout = "This is a custom test command";
                // let installMissingCmd;
                let testStdout;

                await write(tempDir, PACKAGE_JSON, outdatedPackageContents);
                // This time, we install all dependencies
                await exec(
                    tempDir,
                    "yarn add updtr-test-module-1@1.0.0 updtr-test-module-2@2.0.0"
                );

                const yarnLockContentsBefore = await read(tempDir, YARN_LOCK);
                const installedVersionsBefore = getInstalledVersions(
                    tempDir,
                    filterUpdtrTestModule
                );
                const updtr = create({
                    cwd: tempDir,
                    use: USE_YARN,
                    exclude: ["updtr-test-module-2"],
                    updateTo: UPDATE_TO_NON_BREAKING,
                    test: `echo ${ expectedTestStdout } && exit 0`,
                    save: SAVE_CARET,
                    // registry,
                });

                // updtr.on("init/install-missing", ({ cmd }) => {
                //     installMissingCmd = cmd;
                // });
                updtr.on("sequential-update/result", ({ stdout }) => {
                    // echo adds an EOL
                    testStdout = stdout.trim();
                });

                const results = await run(updtr);
                const [
                    tempDirPackageContents,
                    installedVersionsAfter,
                    yarnLockContentsAfter,
                ] = await Promise.all([
                    read(tempDir, PACKAGE_JSON),
                    getInstalledVersions(tempDir, filterUpdtrTestModule),
                    read(tempDir, YARN_LOCK),
                ]);

                // expect(installMissingCmd).toContain(registry);
                expect(testStdout).toBe(expectedTestStdout);
                expect(results).toMatchSnapshot(
                    "advanced configuration > yarn > results"
                );
                expect(tempDirPackageContents).not.toBe(
                    outdatedPackageContents
                );
                expect(
                    JSON.parse(tempDirPackageContents).dependencies
                ).toMatchSnapshot(
                    "advanced configuration > yarn > package.json dependencies"
                );
                expect(installedVersionsAfter).not.toEqual(
                    installedVersionsBefore
                );
                expect(installedVersionsAfter.versions).toMatchSnapshot(
                    "advanced configuration > yarn > installed versions"
                );
                expect(yarnLockContentsBefore).not.toBe(yarnLockContentsAfter);
                expect(yarnLockContentsAfter).toMatch(
                    /updtr-test-module-1@\^1\.0\.0:\s*version "1\.1\.1"/
                );
                expect(yarnLockContentsAfter).toMatch(
                    /updtr-test-module-2@2\.0\.0:\s*version "2\.0\.0"/
                );
            });
        });
    });
});
