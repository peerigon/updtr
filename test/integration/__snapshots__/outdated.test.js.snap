// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`advanced configuration > npm > installed versions 1`] = `
Object {
  "updtr-test-module-1": "1.1.1",
  "updtr-test-module-2": "2.0.0",
}
`;

exports[`advanced configuration > npm > package.json dependencies 1`] = `
Object {
  "updtr-test-module-1": "1.1.1",
  "updtr-test-module-2": "^2.0.0",
}
`;

exports[`advanced configuration > npm > results 1`] = `
Array [
  Object {
    "name": "updtr-test-module-1",
    "rollbackTo": "1.0.0",
    "success": true,
    "updateTo": "1.1.1",
  },
]
`;

exports[`advanced configuration > yarn > installed versions 1`] = `
Object {
  "updtr-test-module-1": "1.1.1",
  "updtr-test-module-2": "2.0.0",
}
`;

exports[`advanced configuration > yarn > package.json dependencies 1`] = `
Object {
  "updtr-test-module-1": "^1.1.1",
  "updtr-test-module-2": "2.0.0",
}
`;

exports[`advanced configuration > yarn > results 1`] = `
Array [
  Object {
    "name": "updtr-test-module-1",
    "rollbackTo": "1.0.0",
    "success": true,
    "updateTo": "1.1.1",
  },
]
`;

exports[`default configuration > npm > installed versions 1`] = `
Object {
  "updtr-test-module-1": "2.0.0",
  "updtr-test-module-2": "2.1.1",
}
`;

exports[`default configuration > npm > package.json dependencies 1`] = `
Object {
  "updtr-test-module-1": "^2.0.0",
  "updtr-test-module-2": "^2.0.0",
}
`;

exports[`default configuration > npm > results 1`] = `
Array [
  Object {
    "name": "updtr-test-module-1",
    "rollbackTo": "1.1.1",
    "success": true,
    "updateTo": "2.0.0",
  },
]
`;

exports[`default configuration > yarn > installed versions 1`] = `
Object {
  "updtr-test-module-1": "2.0.0",
  "updtr-test-module-2": "2.1.1",
}
`;

exports[`default configuration > yarn > package.json dependencies 1`] = `
Object {
  "updtr-test-module-1": "2.0.0",
  "updtr-test-module-2": "^2.0.0",
}
`;

exports[`default configuration > yarn > results 1`] = `
Array [
  Object {
    "name": "updtr-test-module-1",
    "rollbackTo": "1.1.1",
    "success": true,
    "updateTo": "2.0.0",
  },
]
`;
